### PR TITLE
feat: init module balance

### PR DESCRIPTION
**Pull request by gc-garcol**
- [X] ticket: https://trello.com/c/CYpwunCu/2-balance-balance-module
- [ ] high level design: https://lucid.app/lucidchart/a1ae91c4-1002-479d-92b7-8d1f200613e1/edit?viewport_loc=-373%2C0%2C3414%2C1418%2C0_0&invitationId=inv_b5198ab2-7c90-489b-b197-2f3a9b10ba5a

Khoang merge

### CODE REVIEW

File: src/balance/src/main/kotlin/com/exchange/balance/configuration/BalanceKafkaConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+              ConsumerConfig.ISOLATION_LEVEL_CONFIG to "read_committed",
+              ConsumerConfig.MAX_POLL_RECORDS_CONFIG to maxPollRecord,
+              ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG to StringDeserializer::class.java,
+              ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG to JsonNodeDeserializer::class.java))
+    }
+  }
+
+  @Bean
+  fun balanceTopic(): NewTopic {
+    return NewTopic("balance", 1, 1)
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/configuration/BalanceKafkaConfiguration.kt:35]: Move topic name to application.yaml

File: src/balance/src/main/kotlin/com/exchange/balance/model/entity/OffsetEntity.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import jakarta.persistence.Column
+import jakarta.persistence.Entity
+import jakarta.persistence.Id
+import jakarta.persistence.Table
+
+@Entity
+@Table(name = "commit_offsets")
+class OffsetEntity() {
+
+  @Id lateinit var id: String
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/model/entity/OffsetEntity.kt:12]: Using UUID

File: src/balance/src/main/kotlin/com/exchange/balance/repository/disk/BalanceBufferRepository.kt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.exchange.balance.repository.disk
+
+import com.exchange.balance.model.entity.BalanceBufferEntity
+import org.springframework.data.jpa.repository.JpaRepository
+
+interface BalanceBufferRepository : JpaRepository<BalanceBufferEntity, String> {}
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/repository/disk/BalanceBufferRepository.kt:6]: Using UUID

File: src/balance/src/main/kotlin/com/exchange/balance/repository/disk/BalanceRepository.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.exchange.balance.repository.disk
+
+import com.exchange.balance.model.entity.BalanceEntity
+import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.data.jpa.repository.Query
+import java.util.stream.Stream
+
+interface BalanceRepository : JpaRepository<BalanceEntity, String> {
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/repository/disk/BalanceRepository.kt:8]: Using UUID

File: src/balance/src/main/kotlin/com/exchange/balance/repository/disk/OffsetRepository.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.exchange.balance.repository.disk
+
+import com.exchange.balance.model.entity.OffsetEntity
+import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.data.jpa.repository.Modifying
+import org.springframework.data.jpa.repository.Query
+
+interface OffsetRepository : JpaRepository<OffsetEntity, String> {
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/repository/disk/OffsetRepository.kt:8]: Using UUID

Comment (gc-garcol) [reply src/balance/src/main/kotlin/com/exchange/balance/repository/disk/OffsetRepository.kt:8]: no change, it is a String identity for config. Example: id = 'BALANCE_OFFSET'

File: src/balance/src/main/kotlin/com/exchange/balance/model/entity/OffsetEntity.kt (hunk 2 above)
Comment (gc-garcol) [reply src/balance/src/main/kotlin/com/exchange/balance/model/entity/OffsetEntity.kt:12]: it is a truly String type

File: src/balance/src/main/kotlin/com/exchange/balance/service/BalanceService.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+          userId = UUID.fromString(event.userId)
+          amount = BigDecimal.ZERO
+        }
+    balanceInMemoryRepository.upsert(balance)
+    return balance
+  }
+
+  fun deposit(event: DepositBalanceEvent): BalanceEntity? {
+    val balance = balanceInMemoryRepository.findByUserId(UUID.fromString(event.userId))
+    balance?.let {
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/service/BalanceService.kt:48]: use takeif

File: src/balance/src/main/kotlin/com/exchange/balance/service/BalanceService.kt [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      it.amount = it.amount.plus(event.amount)
+      balanceInMemoryRepository.upsert(it)
+      return it
+    }
+    return null
+  }
+
+  fun withdraw(event: WithdrawBalanceEvent): BalanceEntity? {
+    val balance = balanceInMemoryRepository.findByUserId(UUID.fromString(event.userId))
+    balance?.let {
```
Comment (uuhnaut69) [inline src/balance/src/main/kotlin/com/exchange/balance/service/BalanceService.kt:58]: use takeif

File: src/balance/src/main/resources/application-database.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+---
+spring:
+  datasource:
+    url: jdbc:postgresql://localhost:5432/balance
```
Comment (uuhnaut69) [inline src/balance/src/main/resources/application-database.yaml:4]: add env options

**Review by uuhnaut69 — APPROVED** [review]

**Pull request merged**